### ISSUE TITLE
Unclear changelog behaviour under v9 of python-semantic-release

### ISSUE THREAD

**Issue by butler54** [oscal-compass/compliance-trestle#1590]
## Describe the bug

Python semantic release introduced recursive parsing of commits for change logs. 

At the moment the default behaviour of the team is to use squash merges into develop and pull merges into main. The original intent was 'one commit per PR' pulled into develop such that the top line message is clean.

These top level commits are preserved, however, the default behavior of GitHub when squashing commits is to leave the squashed commits in the body of the text. The result is v9 of python-semantic-version is able to find all of the 'squashed commit' messages making it non-sensical

This open issue: https://github.com/python-semantic-release/python-semantic-release/issues/733 appears to fix our problem, however, has been open for over 6 months at this stage.

## To Reproduce

Compare v3.1.0 Changelog vs v3.0.2

## Expected behavior

Reproduce a change log at a higher level.

## Screenshots / Logs.

- [python semantic version issue](https://github.com/python-semantic-release/python-semantic-release/issues/733)
- [v3.1.0 changelog](https://github.com/oscal-compass/compliance-trestle/blob/v3.1.0/CHANGELOG.md)
- [v.3.0.1 changelog](https://github.com/oscal-compass/compliance-trestle/blob/v3.0.1/CHANGELOG.md)

## Environment

- Current release pipeline.

**Comment by butler54** [comment]
There are two parts to fixing this:
1. Correcting the change log
2. Future proofing the change log for this

## Correcting the change log
There are three viable options
1. Wait for the upstream fix.
2. Revert back to python sem-ver 7. 
3. Fork the current angular parser as a custom parser (https://python-semantic-release.readthedocs.io/en/latest/commit-parsing.html#commit-parser-writing-your-own-parser) to fix the problem and (potentially) contribute back


## Enhancements moving forward
A legacy issue really is that our commit log is pretty junky if you get down to the body of the commit (picking a large commit to look at: https://github.com/oscal-compass/compliance-trestle/commit/82caf5fe08796e10532410299d2032b2be7e7d61).

There is no automated way to fix this. To clean this up are options are:
1. Ensure every commit makes sense and merge them all (high risk as no one will look at the commit messages)
2. Ask that every PR is rebased down to a single commit such that authors explicitly have control (I think this would be enforceable).
3. Ask authors to rewrite the squash-merge commit and not just click through (easiest, however, entirely behaviourly driven).

**Comment by jpower432** [comment]
@butler54 Thanks for adding ideas for enhancement moving forward. I have thought about this from a pull request review perspective as well. I wonder if a precursor to the enhancements would be capture PRs sizes or logical chunks and start enforcing norms around that. I think having multiple commits (though it could introduce risk) can make larger PRs easier to review if they are inline with logical steps. From my perspective, squashing down to single commit would make sense for PRs that addresses only one issue.

**Comment by butler54** [comment]
I agree on the size of PRs. it seemed like some where getting quite large recently afaik. The original intent of squash merges was to keep the commit log as 'clean' as possible on the assumption single purpose PRs were being made. Big changes may need some kind of staging branch to manage this in the future.

In terms of cleaning up the commit messages themselves which hit `develop` / `main` there is also a bit of an easier fix here which is changing the behaviour of github on a squahs merge which I forgot about (see below)

This would leave the expectation that the commit author leaves a clean PR title / body and we don't have to wade through all the commits they may have made for fixing ci details if not required.

![Screenshot 2024-06-24 at 10 39 36 AM](https://github.com/oscal-compass/compliance-trestle/assets/19358775/4c03920c-1640-4571-8b78-605bd3e670c9)

**Comment by jpower432** [comment]
@AleJo2995 @butler54 Of the currently defined approaches, I am most in favor of contributing back to the upstream project we currently use.

From my perspective, moving back to v7 for an indefinite period of time would be something I would avoid given that security and bug fixes don't appear to be backported to that version.

Wanted to share another tool that could support automated semantic releases - https://github.com/googleapis/release-please-action. It works with a `squash-merge` workflow and only parsers the top-level commit by default and also supports multiple [commits](https://github.com/googleapis/release-please/tree/ace2bd5dc778f83c33ad5dee6807db5d0afdba36?tab=readme-ov-file#what-if-my-pr-contains-multiple-fixes-or-features) as well. Thought it be worth taking a look as another automated option.

**Comment by AleJo2995** [comment]
@jpower432 can we move this to the next milestone to be fixed? It seems to me needs a little bit of more discussion/investigation and don't want to extend the current milestone anymore. I compromise myself to either contribute back to the repo or to support release-please action.

**Comment by jpower432** [comment]
> @jpower432 can we move this to the next milestone to be fixed? It seems to me needs a little bit of more discussion/investigation and don't want to extend the current milestone anymore. I compromise myself to either contribute back to the repo or to support release-please action. 

Yes I agree @AleJo2995. This seems like it could use more discussion.

**Comment by butler54** [comment]
Looks like there is some movement on this issue: https://github.com/python-semantic-release/python-semantic-release/issues/733#issuecomment-2377871370 something to keep track of.

Also thinking this through an alternative option would be to have the mergebot write the commit to override this.
